src/bitcoin-miner.cpp

```diff
@@ -0,0 +1,723 @@
+    std::vector<unsigned char> vec = ParseHex(tmpstr);
+    std::reverse(vec.begin(), vec.end()); // sent reversed
+    blockheader.hashPrevBlock = uint256(vec);
+
+    // nTime:
+    blockheader.nTime = params["time"].get_int();
+
+    // nBits
+    {
+        std::stringstream ss;
+        ss << std::hex << params["nBits"].get_str();
+        ss >> blockheader.nBits;
+    }
+
+    return blockheader;
+}
+
+
+static void CalculateNextMerkleRoot(uint256 &merkle_root, uint256 &merkle_branch)
+{
+    // Append a branch to the root. Double SHA256 the whole thing:
+    uint256 hash;
+    CHash256()
+        .Write(merkle_root.begin(), merkle_root.size())
+        .Write(merkle_branch.begin(), merkle_branch.size())
+        .Finalize(hash.begin());
+    merkle_root = hash;
+}
+
+static uint256 CalculateMerkleRoot(uint256 &coinbase_hash, std::vector<uint256> merklebranches)
+{
+    uint256 merkle_root = coinbase_hash;
+    for (unsigned int i = 0; i < merklebranches.size(); i++)
+    {
+        CalculateNextMerkleRoot(merkle_root, merklebranches[i]);
+    }
+    return merkle_root;
+}
+
+static bool CpuMineBlockHasher(CBlockHeader *pblock,
+    vector<unsigned char> &coinbaseBytes,
+    std::vector<uint256> merklebranches)
+{
+    uint32_t nExtraNonce = std::rand();
+    uint32_t nNonce = pblock->nNonce;
+    arith_uint256 hashTarget = arith_uint256().SetCompact(pblock->nBits);
+    bool found = false;
+    int ntries = 10;
+    unsigned char *pbytes = (unsigned char *)&coinbaseBytes[0];
+
+    while (!found)
+    {
+        // hashMerkleRoot:
+        {
+            ++nExtraNonce;
+            // 48 - next in arr after Height. (Height in coinbase required for block.version=2):
+            *(unsigned int *)(pbytes + 48) = nExtraNonce;
+            uint256 hash;
+            CHash256().Write(pbytes, coinbaseBytes.size()).Finalize(hash.begin());
+
+            pblock->hashMerkleRoot = CalculateMerkleRoot(hash, merklebranches);
+        }
+
+        //
+        // Search
+        //
+        uint256 hash;
+        while (!found)
+        {
+            // Check if something found
+            if (ScanHash(pblock, nNonce, &hash))
+            {
+                if (UintToArith256(hash) <= hashTarget)
+                {
+                    // Found a solution
+                    pblock->nNonce = nNonce;
+                    found = true;
+                    printf("proof-of-work found  \n  hash: %s  \ntarget: %s\n", hash.GetHex().c_str(),
+                        hashTarget.GetHex().c_str());
+                    break;
+                }
+                else
+                {
+                    if (ntries-- < 1)
+                    {
+                        pblock->nNonce = nNonce; // report the last nonce checked for accounting
+                        return false; // Give up leave
+                    }
+                }
+            }
+        }
+    }
+
+    return found;
+}
+
+double GetDifficulty(uint64_t nBits)
+{
```

Comment:
GetDifficulty() is also a function in Blockchain.cpp.  Maybe this one should be made "static"?